crates/exex/src/db.rs

```diff
@@ -0,0 +1,192 @@
+    ) -> eyre::Result<()> {
+        // Acquire a database connection and begin a transaction.
+        let mut connection = self.connection();
+        let tx = connection.transaction()?;
+
+        // Insert the block into the `block` table.
+        tx.execute(
+            "INSERT INTO block (number, data) VALUES (?, ?)",
+            (block.header.number.to_string(), serde_json::to_string(block)?),
+        )?;
+
+        // Convert the `BundleState` into plain state changes and reverts.
+        let (changeset, _) = bundle.into_plain_state_and_reverts(OriginalValuesKnown::Yes);
+
+        // Process and insert or update account information based on the changeset.
+        for (address, account) in changeset.accounts {
+            if let Some(account) = account {
+                // Insert or update the account in the `account` table.
+                tx.execute(
+                "INSERT INTO account (address, data) VALUES (?, ?) ON CONFLICT(address) DO UPDATE SET data = excluded.data",
+                (address.to_string(), serde_json::to_string(&account)?))?;
+            } else {
+                // Delete the account from the `account` table if it was removed.
+                tx.execute("DELETE FROM account WHERE address = ?", (address.to_string(),))?;
+            }
+        }
+
+        // Commit the transaction to persist all changes.
+        tx.commit()?;
+
+        Ok(())
+    }
+
+    /// Retrieves a block from the database using its block number.
+    ///
+    /// This function queries the database for a block with the specified block number.
+    /// If the block is found, it is deserialized from its JSON representation into a
+    /// `SealedBlockWithSenders` struct. If the block is not found, `None` is returned.
+    pub fn get_block(&self, number: U256) -> eyre::Result<Option<SealedBlockWithSenders>> {
+        // Executes a SQL query to select the block data as a JSON string based on the block number.
+        let block = self.connection().query_row::<String, _, _>(
+            "SELECT data FROM block WHERE number = ?",
+            (number.to_string(),),
+            // Retrieves the first column of the result row as a string.
+            |row| row.get(0),
+        );
+
+        match block {
+            // If the block is found, deserialize the JSON string into `SealedBlockWithSenders`.
+            Ok(data) => Ok(Some(serde_json::from_str(&data)?)),
+            // If no rows are returned by the query, it means the block does not exist in the
+            // database.
+            Err(rusqlite::Error::QueryReturnedNoRows) => Ok(None),
+            // If any other error occurs, convert it into an eyre error and return.
+            Err(e) => Err(e.into()),
+        }
+    }
+
+    /// Inserts a new account if it doesn't exist or updates it if it does.
+    pub fn upsert_account(
+        &self,
+        address: Address,
+        f: impl FnOnce(Option<AccountInfo>) -> eyre::Result<AccountInfo>,
```

Comment:
this seems overly complicated, can't the function just take `address: Address` and `account_info: AccountInfo`?

Reply:
yep just fixed